src/app.py

```diff
@@ -78,90 +81,312 @@ def index():
+- Failure to specify a Content-Type header of application/json for the Request
+- A body in the Request which is valid JSON but is empty
+- Keys which are too long
+- Keys which contain whitespace or other disallowed characters
+
+HTTP 500 Response
+An unexpected error in the server, including unexpected problems with the data store.
 """
 @app.route(rule='/user/keys/<key>', methods=["PUT"])
 @secured(has_write=True)
-def upsert_key_value(key):
-    global worker
+def upsert_key_value(key: Annotated[str, 50]):
+    global ukv_worker
 
-    # Make sure the key is valid before passing it on to upsert operations
-    resp = worker.validateKey(a_key=key)
-    if isinstance(resp, Response):
-        return resp
+    # Make sure the key is valid before passing it on to a database query
+    try:
+        ukv_worker.validate_key(a_key=key)
+    except ukvEx.UKVKeyFormatException as e_400:
+        return jsonify({'error': e_400.message}), 400
 
     try:
-        resp = worker.upsertKeyValue(req=request
-                                     , valid_key=key)
-        return resp
-    except Exception as e:
-        logger.error(e, exc_info=True)
-        return make_response(f"Unexpected error setting key '{key}'. See logs."
-                             , 500)
+        success_msg = ukv_worker.upsert_key_value(req=request
+                                                  , valid_key=key)
+        return jsonify({'message': success_msg})
+    except (ukvEx.UKVValueFormatException, ukvEx.UKVRequestFormatException) as e_400:
+        return jsonify({'error': e_400.message}), 400
+    except (ukvEx.UKVDataStoreQueryException, Exception) as e_500:
+        msg = f"Unexpected error setting key '{key}'."
+        logger.exception(e_500)
+        return jsonify({'error': f"{msg} See logs."}), 500
 
 """
 An endpoint to retrieve the value matching the given key for the authenticated user.
-On success returns the JSON value for the user's key
 
 Parameters
 ----------
-key : The user's key to be retrieved, as specified on the API endpoint 
+key : The name of the user's key for which to retrieve the value. 
 
 Returns
 -------
-Valid JSON stored under the user's key
+HTTP 200 Response with Content-Type application/json
+The JSON stored as the value for the given key.
+
+HTTP 400 Response with Content-Type application/json
+The JSON body is a dictionary with an error message describing the problem that kept
+key/value pair from being retrieved.  Examples causes are
+- Keys which are too long
+- Keys which contain whitespace or other disallowed characters
+
+HTTP 404 Response with Content-Type application/json
+The JSON body is a dictionary with an error message indicating the key was not found in the data store.
+
+HTTP 500 Response
+An unexpected error in the server, including unexpected problems with the data store.
 """
 @app.route(rule='/user/keys/<key>', methods=["GET"])
-def get_key_value(key):
-    global worker
+def get_key_value(key: Annotated[str, 50]):
```

Comment:
To avoid confusion, does `Annotated[str, 50]` suggest the max length is 50 or the key length needs to be exact 50?

Reply:
It is just type hinting introduced in Python 3.10 for readability, documentation, and static code analysis. Not "enforced", so the explicit verification of length remains in the code. When we were at Python 3.9, I would have just used `str` for the hint.